Inductor cpp wrapper: support MkldnnRnnLayer

chunyuan-w opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107858

1. Directly use the `codegen` function of the parent class which already supported both python and cpp wrapper.
2. The output of the `at::mkldnn_rnn_layer` OP is actually a `std::tuple` https://github.com/pytorch/pytorch/blob/1491bae277668fac459937c874a49c3bb8adedcb/aten/src/ATen/native/mkldnn/RNN.cpp#L218 Fix the type when calling `MultiOutput`.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @ngimel

chunyuan-w committed:
Inductor cpp wrapper: support MkldnnRnnLayer

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: support MkldnnRnnLayer"


1. Directly use the `codegen` function of the parent class which already supported both python and cpp wrapper.
2. The output of the `at::mkldnn_rnn_layer` OP is actually a `std::tuple` https://github.com/pytorch/pytorch/blob/1491bae277668fac459937c874a49c3bb8adedcb/aten/src/ATen/native/mkldnn/RNN.cpp#L218 Fix the type when calling `MultiOutput`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

chunyuan-w committed:
Update on "Inductor cpp wrapper: support MkldnnRnnLayer"


1. Directly use the `codegen` function of the parent class which already supported both python and cpp wrapper.
2. The output of the `at::mkldnn_rnn_layer` OP is actually a `std::tuple` https://github.com/pytorch/pytorch/blob/1491bae277668fac459937c874a49c3bb8adedcb/aten/src/ATen/native/mkldnn/RNN.cpp#L218 Fix the type when calling `MultiOutput`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

chunyuan-w commented on test/inductor/test_torchinductor.py:
```diff
@@ -221,8 +221,9 @@ def clone_preserve_strides(x, device=None):
 
-@patch.object(config, "debug", True)
 def run_and_get_cpp_code(fn, *args, **kwargs):
+    old_debug_config = config.debug
```
`run_and_get_cpp_code` is called twice when running `python -u test/inductor/test_cpp_wrapper.py -k test_lstm_packed_change_input_sizes`.
When using `patch` from `unittest.mock`, exiting the decorator twice will result in: `AttributeError: '_patch' object has no attribute 'is_local'`, because `mock._patch.__exit__` is executed several times on the same `_patch` object and since `is_local` has been deleteded during the first exit, the second call fails.

jgong5 replied:
Suggest to add the note as code comment.

chunyuan-w committed:
Update on "Inductor cpp wrapper: support MkldnnRnnLayer"


1. Directly use the `codegen` function of the parent class which already supported both python and cpp wrapper.
2. The output of the `at::mkldnn_rnn_layer` OP is actually a `std::tuple` https://github.com/pytorch/pytorch/blob/1491bae277668fac459937c874a49c3bb8adedcb/aten/src/ATen/native/mkldnn/RNN.cpp#L218 Fix the type when calling `MultiOutput`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

jgong5 commented on test/inductor/test_torchinductor.py:
```diff
@@ -238,6 +239,7 @@ def run_and_get_cpp_code(fn, *args, **kwargs):
     s = log_capture_string.getvalue()
     output_code_log.setLevel(prev_level)
     output_code_log.removeHandler(ch)
+    config.debug = old_debug_config
```
safer to do "try ... finally" here to make sure `config.debug` is restored?

chunyuan-w replied:
I found that using the `patch` context manager instead of the decorator could fix this issue.
I've refined the code and added comments.

chunyuan-w committed:
Update on "Inductor cpp wrapper: support MkldnnRnnLayer"


1. Directly use the `codegen` function of the parent class which already supported both python and cpp wrapper.
2. The output of the `at::mkldnn_rnn_layer` OP is actually a `std::tuple` https://github.com/pytorch/pytorch/blob/1491bae277668fac459937c874a49c3bb8adedcb/aten/src/ATen/native/mkldnn/RNN.cpp#L218 Fix the type when calling `MultiOutput`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

jgong5 commented on test/inductor/test_torchinductor.py:
```diff
@@ -222,24 +222,25 @@ def clone_preserve_strides(x, device=None):
-    output_code_log.removeHandler(ch)
-    config.debug = old_debug_config
+    # We use the patch context manager instead of using it as a decorator.
+    # In this way, we can ensure that the attribute is patched and unpacted correctly
```
nit: typo
```suggestion
    # In this way, we can ensure that the attribute is patched and unpatched correctly
```

chunyuan-w replied:
Fixed the typo.

chunyuan-w committed:
Update on "Inductor cpp wrapper: support MkldnnRnnLayer"


1. Directly use the `codegen` function of the parent class which already supported both python and cpp wrapper.
2. The output of the `at::mkldnn_rnn_layer` OP is actually a `std::tuple` https://github.com/pytorch/pytorch/blob/1491bae277668fac459937c874a49c3bb8adedcb/aten/src/ATen/native/mkldnn/RNN.cpp#L218 Fix the type when calling `MultiOutput`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

chunyuan-w commented:
@pytorchbot merge
